src/main/java/org/opensearch/security/configuration/SecurityConfigVersionHandler.java

```diff
@@ -0,0 +1,337 @@
+            if (latestVersion == null || latestVersion.getVersion_id() == null || !latestVersion.getVersion_id().startsWith("v")) {
+                return "v1";
+            }
+            int currentVersionNumber = Integer.parseInt(latestVersion.getVersion_id().substring(1));
+            return "v" + (currentVersionNumber + 1);
+        } catch (Exception e) {
+            log.error("Failed to fetch latest version from {}", SecurityConfigVersionsIndex, e);
+            throw new RuntimeException("Failed to fetch next version id", e);
+        }
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public SecurityConfigVersionDocument.Version<?> buildVersionFromSecurityIndex(String versionId, String modified_by) throws IOException {
+        Instant now = Instant.now();
+        String timestamp = now.toString();
+
+        SecurityConfigVersionDocument.Version<?> version = new SecurityConfigVersionDocument.Version<>(
+            versionId,
+            timestamp,
+            new HashMap<>(),
+            modified_by
+        );
+
+        ConfigurationMap allConfigs = cr.getConfigurationsFromIndex(CType.values(), false);
+
+        for (CType<?> cType : CType.values()) {
+            SecurityDynamicConfiguration<?> dynamicConfig = allConfigs.get(cType);
+
+            if (dynamicConfig == null || dynamicConfig.getCEntries() == null || dynamicConfig.getCEntries().isEmpty()) {
+                version.addSecurityConfig(
+                    cType.toLCString(),
+                    new SecurityConfigVersionDocument.SecurityConfig<>(timestamp, new HashMap<>())
+                );
+            } else {
+                version.addSecurityConfig(
+                    cType.toLCString(),
+                    new SecurityConfigVersionDocument.SecurityConfig(timestamp, new TreeMap<>(dynamicConfig.getCEntries()))
+                );
+            }
+        }
+
+        return version;
+    }
+
+    public <T> void saveCurrentVersionToSystemIndex(SecurityConfigVersionDocument.Version<T> version) {
+        try {
+            SecurityConfigVersionDocument document = configVersionsLoader.loadFullDocument();
+            if (shouldSkipVersionUpdate(document, version)) {
+                return;
+            }
+            // Otherwise, add the new version and update the document
+            document.addVersion(version);
+            writeSecurityConfigVersion(document, document.getSeqNo(), document.getPrimaryTerm());
+
+            log.info("Successfully saved version {} to {}", version.getVersion_id(), SecurityConfigVersionsIndex);
+
+            // Async retention task
+            threadPool.generic().submit(() -> {
+                try {
+                    applySecurityConfigVersionIndexRetentionPolicy();
+                } catch (Exception e) {
+                    log.warn("Retention policy async failed", e);
+                }
```

Comment:
looks like ``saveCurrentVersionToSystemIndex`` will first store the new version and later try and remove older versions if number of versions exceed. While this might be generally okay, this currently means we temporarily allow more versions than configured by an user. It might leave the system in an inconsistent state if the retention enforcement thread fails.